draft-ietf-jsonpath-base.md

```diff
@@ -413,51 +433,70 @@ the grammar in this document, its UTF-8 form SHOULD first be decoded into
 Unicode code points as described
 in {{RFC3629}}.
 
-
-## Terminology
-
-A JSON value is logically a tree of nodes.
-
-Each node holds a JSON value (as defined by {{RFC8259}}) of one of the
-types object, array, number, string, or one of the literals `true`,
-`false`, or `null`.
-The type of the JSON value held by a node is
-sometimes referred to as the type of the node.
-
-
-## Implementation
-
-An implementation of this specification, from now on referred to simply as
-"an implementation", SHOULD takes two inputs, a JSONPath and a JSON value,
-and produce
-a possibly empty list of nodes of the JSON value which are selected by
-the JSONPath or an error (but not both).
-
-If no node is selected and no error has occurred, an implementation MUST
-return an empty list of nodes.
-
-Syntax errors in the JSONPath SHOULD be detected before selection is attempted
-since these errors do not depend on the JSON value.
-Therefore, an implementation SHOULD take a JSONPath and produce an optional
-syntax error and then,
-if and only if an error was not produced, SHOULD take a JSON value and
-produce a list of nodes or an error (but not both).
-
-Alternatively, an implementation MAY take a JSONPath and a JSON value
-and produce a list of nodes or an optional error (but not both).
-
-For any implementation, if a syntactically invalid JSONPath is provided,
-the implementation MUST return an error.
-
-If a syntactially invalid JSON value is provided, any implementation SHOULD
-return an error.
+A string to be used as a JSONPath query needs to be *well-formed* and
+*valid*.
+A string is a well-formed JSONPath query if it conforms to the syntax
+of JSONPath.
+A well-formed JSONPath query is valid if it also fulfills all semantic
+requirements posed by this document.
+
+Well-formedness and validity of JSONPath queries are independent of
+the JSON value the query is applied to; no further errors can be
+raised during application of the query to a JSON value.
+
+(Obviously, an implementation can still fail when executing a JSONPath
+query, e.g., because of resource depletion, but this is not modeled in
+the present specification.)
+
+> Discussion: are we ready for the error model defined here, i.e. one
+> that does not have runtime errors, but only compile time errors?
+
+## Processing Model
+
+A valid JSONPath query is executed against a JSON value, the *Argument*, and
+produces a list of zero or more nodes of the JSON value which are
+selected by the JSONPath.
+(Note that the term JSON value also implies that this value complies
+to the definitions in {{-json}}, i.e., is indeed a JSON value.)
+
+The JSONPath query is a sequence of zero or more *selectors*, each of
+which is applied to the result of the previous selector and leaves
+input to the next selector.
```

Comment:
```suggestion
which is applied to each node in the result of the previous selector and outputs
a single list of nodes to be used as input by the next selector.
```

See [discussion](https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-jsonpath/issues/14#issuecomment-694687029) over representing the algorithm as iterating over the individual results of the previous selector vs taking the full result set as a single input.

The suggestion above is in line with the rest of the content of the PR so far: each selector iterates over the output from the previous selector.

Reply:
The changed text implies that each function invocation of the selected leaves a node list that is then independently fed to the next selector.  I think the discussion resolved to doing the combine1 step first, yielding a single nodelist, and then applying the next selector again to each node in that nodelist.

Reply:
I'm not sure I followed that.  My suggested changes align this statement with the rest of the your descriptions to say that each selector takes the nodelist (combined) output by the previous selector as a single argument and iterates over the nodes internally.

The other form is to say that the selector is fed each node individually by the processor, which doesn't line up with the rest of the PR.

Reply:
The processing model actually says that the selector processing the input nodelist into the output nodelist by processing each entry of the input nodelist into a partial output nodelist and then uses combine1 to put these together into the output nodelist.  This probably needs better terminology.  Your text tells me that a selector is applied to each element of the input nodelist and does not say what we do with all those partial output nodelists.

Reply:
After rereading my comments, I see the confusion between what I've been saying and my addition of "each node in."  I think this can be removed from my suggestion:

```suggestion
which is applied to the result of the previous selector and outputs
a new nodelist to be used as input by the next selector.
```

Regarding how the results of each node are combined, this paragraph doesn't address it at all.  This paragraph instead focuses on the input/output stream of the selectors.  The "combining" logic is something internal to how the selectors work, which is discussed later when you're talking about the "combine1" function.  I think this is a good layout for the spec as it stands in this proposal.

Reply:
The word nodelist is first discussed in the next sentence, so this is simply pulling in an unneeded forward reference.

Reply:
Still, "leaves input to..." is odd wording.

Reply:
89d7d72: leaves ➔ provides